File: leetcode/Python/populating-next-right-pointers-in-each-node-ii.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        dq = collections.deque()
        sz, cur, prev = 0, None, None
        if root is None:
            return root
        dq.append(root)
        while dq:
            prev = None
            for i in range(len(dq)):
                cur = dq.popleft()
                if cur.left is not None:
                    dq.append(cur.left)
                if cur.right is not None:
                    dq.append(cur.right)
                if prev is not None:
                    prev.next = cur
                prev = cur
        return root

File: leetcode/C++/populating-next-right-pointers-in-each-node-ii.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        int sz = 0;
        Node* prev=nullptr, *cur=nullptr;
        queue<Node*> q;
        if(root==nullptr) return root;
        q.push(root);
        while(!q.empty()){
            prev=nullptr;
            sz = q.size();
            for(int i=0; i<sz; ++i){
                cur=q.front(); q.pop();
                if(cur->left) q.push(cur->left);
                if(cur->right) q.push(cur->right);
                if(prev) prev->next=cur;
                prev=cur;
            }
        }
        return root;
    }
};